### ISSUE TITLE
Ignore ProjectCacheDir when snapshotting

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#5941]
I tried to copy a file to the project directory with
```groovy
task updateReadme(type: Copy) {
    inputs.property 'version', version
    from '.github/readme_template.md'
    into '.'
    rename { 'README.md' }
    expand version: version
}
```
but this fails when Gradle tries to calculate the output files snapshot.
When it tries to calculate the snapshot for `./.gradle/4.8.1/fileHashes/fileHashes.lock` it fails with

```exception
org.gradle.api.UncheckedIOException: Failed to capture snapshot of output files for task ':updateReadme' property 'destinationDir' during up-to-date check.
	at org.gradle.api.internal.changedetection.state.CacheBackedTaskHistoryRepository.snapshotTaskFiles(CacheBackedTaskHistoryRepository.java:336)
	at org.gradle.api.internal.changedetection.state.CacheBackedTaskHistoryRepository.createExecution(CacheBackedTaskHistoryRepository.java:158)
	at org.gradle.api.internal.changedetection.state.CacheBackedTaskHistoryRepository.access$100(CacheBackedTaskHistoryRepository.java:66)
	at org.gradle.api.internal.changedetection.state.CacheBackedTaskHistoryRepository$1.getCurrentExecution(CacheBackedTaskHistoryRepository.java:116)
	at org.gradle.api.internal.changedetection.changes.DefaultTaskArtifactStateRepository$TaskArtifactStateImpl.calculateCacheKey(DefaultTaskArtifactStateRepository.java:129)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.doResolve(ResolveBuildCacheKeyExecuter.java:106)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.access$000(ResolveBuildCacheKeyExecuter.java:48)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter$1.run(ResolveBuildCacheKeyExecuter.java:90)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.resolve(ResolveBuildCacheKeyExecuter.java:87)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.execute(ResolveBuildCacheKeyExecuter.java:65)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
	at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$ExecuteTaskAction.execute(DefaultTaskExecutionGraph.java:262)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$ExecuteTaskAction.execute(DefaultTaskExecutionGraph.java:246)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:136)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.execute(DefaultTaskPlanExecutor.java:201)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.executeWithTask(DefaultTaskPlanExecutor.java:192)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.gradle.api.UncheckedIOException: Failed to create MD5 hash for file content.
	at org.gradle.internal.hash.DefaultStreamHasher.hash(DefaultStreamHasher.java:40)
	at org.gradle.internal.hash.DefaultFileHasher.hash(DefaultFileHasher.java:40)
	at org.gradle.api.internal.changedetection.state.CachingFileHasher.snapshot(CachingFileHasher.java:88)
	at org.gradle.api.internal.changedetection.state.CachingFileHasher.snapshot(CachingFileHasher.java:75)
	at org.gradle.api.internal.changedetection.state.CachingFileHasher.hash(CachingFileHasher.java:61)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter.fileSnapshot(DefaultFileSystemSnapshotter.java:244)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter.access$700(DefaultFileSystemSnapshotter.java:60)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter$FileVisitorImpl.visitFile(DefaultFileSystemSnapshotter.java:297)
	at org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker$1.visitFile(Jdk7DirectoryWalker.java:86)
	at org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker$1.visitFile(Jdk7DirectoryWalker.java:59)
	at java.nio.file.Files.walkFileTree(Files.java:2670)
	at org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker.walkDir(Jdk7DirectoryWalker.java:59)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.walkDir(DirectoryFileTree.java:154)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.visitFrom(DirectoryFileTree.java:132)
	at org.gradle.api.internal.file.collections.DirectoryFileTree.visit(DirectoryFileTree.java:117)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter.snapshotAndCache(DefaultFileSystemSnapshotter.java:192)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter.access$500(DefaultFileSystemSnapshotter.java:60)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter$4.create(DefaultFileSystemSnapshotter.java:174)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter$4.create(DefaultFileSystemSnapshotter.java:169)
	at org.gradle.cache.internal.ProducerGuard$StripedProducerGuard.guardByKey(ProducerGuard.java:115)
	at org.gradle.api.internal.changedetection.state.DefaultFileSystemSnapshotter.snapshotDirectoryTree(DefaultFileSystemSnapshotter.java:169)
	at org.gradle.api.internal.changedetection.state.AbstractFileCollectionSnapshotter$FileCollectionVisitorImpl.visitDirectoryTree(AbstractFileCollectionSnapshotter.java:100)
	at org.gradle.api.internal.changedetection.state.AbstractFileCollectionSnapshotter$FileCollectionVisitorImpl.visitCollection(AbstractFileCollectionSnapshotter.java:84)
	at org.gradle.api.internal.file.AbstractFileCollection.visitRootElements(AbstractFileCollection.java:282)
	at org.gradle.api.internal.file.CompositeFileCollection.visitRootElements(CompositeFileCollection.java:206)
	at org.gradle.api.internal.changedetection.state.AbstractFileCollectionSnapshotter.snapshot(AbstractFileCollectionSnapshotter.java:55)
	at org.gradle.api.internal.changedetection.state.DefaultGenericFileCollectionSnapshotter.snapshot(DefaultGenericFileCollectionSnapshotter.java:38)
	at org.gradle.api.internal.changedetection.state.CacheBackedTaskHistoryRepository.snapshotTaskFiles(CacheBackedTaskHistoryRepository.java:334)
	... 43 more
Caused by: java.io.IOException: Der Prozess kann nicht auf die Datei zugreifen, da ein anderer Prozess einen Teil der Datei gesperrt hat
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:233)
	at org.gradle.internal.hash.DefaultStreamHasher.doHash(DefaultStreamHasher.java:54)
	at org.gradle.internal.hash.DefaultStreamHasher.hash(DefaultStreamHasher.java:38)
	... 70 more

```

**Comment by Vampire** [comment]
Workaround:
```groovy
task updateReadme {
    inputs.property 'version', version
    inputs.file '.github/readme_template.md' withPropertyName 'readmeTemplate'
    outputs.file 'README.md' withPropertyName 'readme'

    doLast {
        copy {
            from '.github/readme_template.md'
            into '.'
            rename { 'README.md' }
            expand version: version
        }
    }
}
```

**Comment by wolfs** [comment]
Thank you for reporting. Is this a duplicate of https://github.com/gradle/gradle/issues/4663?

**Comment by Vampire** [comment]
Yes, actually this seems to be exactly the same thing.
But sad that it wont get fixed. :-(
There are use-cases that are valid, even if just a few.
And the suspicion in that ticket that another process locked the file was wrong, I checked with Process Explorer and the only process holding the file hostage was the Gradle daemon itself.

Actually, why do you consider all files in the output directory as output .
I would see two possible solutions (that could also both be implemented) that would also improve speed and reduce false out-of-date detections.
1. When only copying one file, allow or require to specify an output file instead of an output directory, that would also do a rename implcitly without the need to define the rename explicitly.
2. Generally calculate the concrete output files from the input files and not consider the whole output directory. Take the input files, map them to files in the output directory considering things like rename and then only consider these files as output files.

**Comment by wolfs** [comment]
I am not doubting the use-case (copying a file in the project directory). I am still convinced that using the `Copy` task for that use-case is the right thing to do.

Do I understand correctly that you want up-to-date checks and copying a single file to another location? Without up-to-date checks you already have the solution: Use `Project.copy`.

With up-to-date checks your solution 1) sounds to me much like a different task type to `Copy`. `CopyFile` maybe? This task can then declare the output file.

BTW: `rename` actions are not tracked as inputs to the `Copy` task, so changing those will not make the task out of date.

2) would be possible for the `Copy` task, though not completely trivial because of renames. There is also `eachFile`, which may be tricky to track. For other tasks, 2) breaks down completely. We need a solution for tracking output files which works for all kinds of tasks.

We don't consider all the files in the output directory as an output. What we do is that we take a snapshot of the whole output directory before we start executing the task and after we executed the task. In this way we can determine which files have been added/changed by the task so that we know what the actual output of the task is.

**Comment by Vampire** [comment]
> I am still convinced that using the Copy task for that use-case is the right thing to do.

I guess you mean "not", otherwise it would be strange that you close the issue :-D

> Without up-to-date checks you already have the solution: Use `Project.copy`.

Even with up-to-date checks I have a work-around, as shown in my first comment.
But I wouldn't call this a solution.
Gradle is all about convenient and easy to use.
If someone that does not debug Gradle like me, he naively tries it with the `Copy` task, gets an error that is abolutely meaningless to him and maybe abandons Gradle, or reports the bug again.
Even if this bug does not get fixed, there should be a more meaningful error message like `using the cache directory .gradle or a parent directory as destination directory is not supported, move your cache directory using [...] or use `Project.copy` and define your outputs explicitly for up-to-date checks to work` or something similar, so that the problem is instantly clear and the "solution" (I'd still call it work-around :-P) obvious.

> With up-to-date checks your solution 1) sounds to me much like a different task type to Copy. CopyFile maybe? This task can then declare the output file.

Would also be an option, but then please also add a hint during the `Copy` task execution that recommends using `CopyFile` if only one input file is defined generally and this hint must be shown before the output files hashing fails, or if it fails and there is only one input file, then recommend using the `CopyFile` task instead of recommending `Project.copy`.

> BTW: `rename` actions are not tracked as inputs to the Copy task, so changing those will not make the task out of date.

Oh, really? o_O That I think is rather strange and more a bug, it is a significant property and definitly requires the task to re-run, don't you think so?

> We don't consider all the files in the output directory as an output. What we do is that we take a snapshot of the whole output directory before we start executing the task and after we executed the task. In this way we can determine which files have been added/changed by the task so that we know what the actual output of the task is.

Ah, I see and for the output-files changed check you look which output files were there when last executing the task and whether these changed or disappeared. If input files change the task is out-of-date anyway.

Actually what comes to my mind right now is even worse, if one develops the build on Linux or Mac everythig works fine and as soon as some poor Windows user executes the build it blows up, this is also not too nice and could harm the Gradle reputation regarding easy-to-useness.

Maybe you could then just not consider the project cache dir contenst during these snapshots calculations or make them work somehow on Windows too if possible.

**Comment by wolfs** [comment]
> > I am still convinced that using the Copy task for that use-case is the right thing to do.
>
> I guess you mean "not", otherwise it would be strange that you close the issue :-D

Yes, I meant "not". Corrected.

> Even if this bug does not get fixed, there should be a more meaningful error message like `using the cache directory .gradle or a parent directory as destination directory is not supported, move your cache directory using [...] or use Project.copy and define your outputs explicitly for up-to-date checks to work` or something similar, so that the problem is instantly clear and the "solution" (I'd still call it work-around :-P) obvious.

Agreed, the error message should be better.

> Maybe you could then just not consider the project cache dir contenst during these snapshots calculations or make them work somehow on Windows too if possible.

Excluding the `.gradle` directory for snapshotting by default sounds like a good idea. We maybe should do that. FTR, `ProjectCacheDir` carries the information where the directory is.

**Comment by Vampire** [comment]
So, will you reopen this for the things to be changed? Will you open them as new tickets? Do you expect me to report them as new tickets?

**Comment by wolfs** [comment]
@Vampire I created #5972, renamed the current issue and re-opened it. While it would be good the fix this issue or at least the error message, I don't see that high on our priority list right now.

**Comment by Vampire** [comment]
At least it is on the list and I know how to work-around it, so 👍 :-)
